[#2] feat: 회원가입 기능

Yo0oN opened the pull request:
일반 사용자 회원가입 기능 추가.

[#2] 회원가입 기능

Yo0oN committed:
feat: 회원가입 기능

일반 사용자 회원가입 기능 추가.

[#2] 회원가입 기능

Jane096 commented on src/main/java/com/festa/dto/HostDTO.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
MemberDTO부분에는 `@Getter` 만 선언이 되어 있고 HostDTO 쪽에는 `@Getter` , `@Setter` 둘 다 선언되어 있는데 혹시 이유가 있으신가요?

Yo0oN replied:
HostDTO는 처음 설정시 그냥써놨던 부분이라 아직 완성된건 아니고, MemberDTO에 getter만 있는건 회원가입부분에서 아이디 확인중 getter가 필요해서 사용했습니다. 아직 회원가입 기능을 완성한것은 아니라 이후에 setter가 필요하다면 추가할 예정입니다

f-lab-dev replied:
이 부분은 `java immutable`로 검색해보시면 좋을 것 같네요~

f-lab-dev replied:
그리고 서로 이런 리뷰 좋습니다. 이해 안돼는건 물어보고 피드백 줄 수 있는건 최대한 주시면 좋습니다~

f-lab-dev commented on pom.xml:
```diff
@@ -51,6 +51,18 @@
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-autoconfigure</artifactId>
+            <version>2.3.4.RELEASE</version>
```
여기서 버전을 명시적으로 설정해주신 이유가 있으신가요?? 만약 명시하지 않으면 어떻게 될까요?

WIGOyoonj replied:
가장 최신의 버전을 사용하기 위해 적어두었던 것인데 spring-boot-starter-parent가 있으니 버전을 설정해주지 않아도 알아서 다른 라이브러리들과 충돌되지 않는 버전으로 관리가 됩니다.
해당 부분은 삭제하겠습니다.

Yo0oN replied:
가장 최신의 버전을 사용하기 위해 적어두었던 것인데 spring-boot-starter-parent가 있으니 버전을 설정해주지 않아도 알아서 다른 라이브러리들과 충돌되지 않는 버전으로 관리가 됩니다.
해당 부분은 삭제하겠습니다.

f-lab-dev commented on src/main/java/com/festa/controller/HostController.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+
+@Controller
```
`RestController`가 아니게 설정하신 이유가 있으신가요?

Yo0oN replied:
Controller를 사용하면 View를 반환하고, 데이터를 직접 반환하고 싶다면 ResponseBody와 함께 사용해야 했으나, RestConroller는 Controller와 ResponseBody가 합쳐진 것으로 JSON/XML 타입의 HTTP 응답할 수 있습니다.
RestController에 대한 이해가 부족해서 Controller를 사용했으나, 수정하겠습니다.

f-lab-dev commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -0,0 +1,32 @@
+     * @param memberDTO
+     * @return
+     */
+    @PostMapping(value = "/signUpMember")
```
이 부분은 URL에 대한 설계가 필요할 것 같네요. 아래의 문서를 참고해보시면 좋겠습니다~

https://docs.microsoft.com/ko-kr/azure/architecture/best-practices/api-design

f-lab-dev commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -0,0 +1,32 @@
+    @PostMapping(value = "/signUpMember")
+    public HttpStatus signUpMember(@RequestBody @Valid MemberDTO memberDTO){
+        memberService.signUpMember(memberDTO);
+        return HttpStatus.CREATED;
```
여기는 status code를 목적에 맞게 잘 써주셨네요 👍

f-lab-dev commented on src/main/java/com/festa/dao/MemberDAO.java:
```diff
@@ -0,0 +1,23 @@
+@Repository
+public interface MemberDAO {
+    /**
+     * 회원 가입
```
이런 주석들은 `메소드 시그니처`에서 충분히 유추가 가능해보입니다~
꼭 필요한 주석일까요?

Yo0oN replied:
메소드마다 설명이 필요하다고 생각해서 적었던 부분인데 이름으로 유추가능하다면 주석을 달지 않아도 되는건가요?

f-lab-dev replied:
"이름으로 유추가 가능하기 때문에 달지 않는다" 라는 것은 너무 단편적이고요~
모든 클래스와 메소드를 설계할 때에는 최대한 기능이 어떤것인지 유추 가능하게 만들어야합니다.
그리고 그것만으로 파악하기 힘든 내부 로직에 대해 주석으로 남기면 좋고요~

f-lab-dev commented on src/main/java/com/festa/dto/HostDTO.java:
```diff
@@ -0,0 +1,14 @@
+@Setter
+public class HostDTO {
+    String hostname;
+    String password;
```
명확한 접근제어자도 붙여주시면 좋겠습니다~

f-lab-dev commented on src/main/java/com/festa/dto/MemberDTO.java:
```diff
@@ -0,0 +1,28 @@
+import javax.validation.constraints.NotBlank;
+import javax.validation.constraints.Pattern;
+
+@Getter
+public class MemberDTO {
```
여기는 생성자가 없는데 객체 생성은 어떻게 할까요??

Yo0oN replied:
@AllArgsConstructor를 추가하여 모든 멤버변수를 매개변수로 가지는 생성자를 만들도록 수정하였습니다.

f-lab-dev commented on src/main/java/com/festa/service/MemberService.java:
```diff
@@ -0,0 +1,45 @@
+     * @param memberDTO
+     */
+    public void signUpMember(MemberDTO memberDTO) {
+        checkMemberID(memberDTO.getEmail());
```
`checkMemberID`라는 이름에서 역할을 유추할 수 있을까요? `validateMemberID`로 하면 어떨까요?

Yo0oN replied:
다음에는 이름으로도 역할을 알 수 있도록 만들겠습니다. checkMemberID는 validateMemberID나 duplicatedMemberID로 수정하겠습니다.

f-lab-dev commented on src/main/java/com/festa/service/MemberService.java:
```diff
@@ -0,0 +1,45 @@
+    public void checkMemberID(String email) {
+        List<MemberDTO> checkMemberID = memberDAO.checkMemberID(email);
+        if (!checkMemberID.isEmpty()) {
+            throw new IllegalStateException("이미 등록된 이메일입니다.");
```
이것은 사용자가 잘못된 이메일을 입력한 것 같네요. 그러면 `IllegalStateException`가 맞을까요?

Yo0oN replied:
IllegalStateException은 메서드가 부적절한때 호출되는 예외이기 때문에 맞지 않는것같습니다. 대신 예외를 새로 만들거나 아이디가 중복될 경우 false를 리턴해서 그럴경우 상태코드를 412로 응답하도록 바꾸는식으로 하는건 괜찮을까요?

f-lab-dev replied:
여기서는 예외 상황이 맞을 것 같네요~ 회원가입을 할때 이미 등록된 이메일을 가입하는 시나리오는 예외상황이니까요

f-lab-dev commented:
리뷰받은 내용은 수정하셔서 동일한 브랜치에 push해주세요~

Yo0oN committed:
feat: 회원가입 기능

일반 사용자, 주최자 회원가입 Service, DAO, DTO 합침.
jasypt 암호화 추가

[#2] 회원가입 기능

Yo0oN committed:
암호화 관련 설정 수정

Yo0oN committed:
암호화 관련 설정 수정

Jane096 commented on src/main/java/com/festa/controller/HostController.java:
```diff
@@ -1,16 +1,31 @@
-import com.festa.dto.HostDTO;
-import com.festa.service.HostService;
+import com.festa.dto.SignUpDTO;
+import com.festa.service.AccountsService;
```
Controller만 기존의 member / host 로 나누고 service, dao, dto, mapper.xml 모두 통합을 하셨던데 controller만 따로 분리한 이유가 있는지 궁금합니다!

Yo0oN replied:
member와 host는 경로만 다를 뿐 회원관련 기능은 동일하고, 같은 테이블을 사용하기 때문에 통합하였습니다. 이후에 서로 다른 기능을 사용할 경우 해당 기능은 DAO, DTO, mapper.xml을 분리하여 작성할 생각입니다!

f-lab-dev commented on pom.xml:
```diff
@@ -35,11 +36,14 @@
             <optional>true</optional>
         </dependency>
+
+        <!-- lombok 설정 -->
```
아래의 내용에서 lombok 의존성을 설정했다라는 것은 유추가 가능합니다. 어떤 의도로 lombok 의존성을 추가하신건지가 잘 드러나면 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/festa/config/DatabaseConfig.java:
```diff
@@ -0,0 +1,48 @@
+
+    @Bean
+    public DataSource dataSource(HikariConfig hikariConfig) { // HikariConfig Bean 주입
+        //return new HikariDataSource(hikariConfig);
```
push하는 것은 실서버 소스에 합칠 소스를 push하는 것입니다. 이런 주석이 쌓이면 읽기 힘든 코드가 되어버리니 이런 필요없는 코드는 삭제하고 push해주세요~

Yo0oN replied:
네 주의하겠습니다

f-lab-dev commented on src/main/java/com/festa/config/DatabaseConfig.java:
```diff
@@ -0,0 +1,48 @@
+    @Bean
+    @ConfigurationProperties(prefix = "spring.datasource.hikari")
+    public HikariConfig hikariConfig() { // application.properties에 작성된 설정
+        return new HikariConfig();
```
이 bean을 선언하지 않고 properties 파일 설정만으로도 DB설정이 가능할 것 같은데요. 따로 bean을 정의해주신 이유가 있으실까요?

f-lab-dev commented on src/main/java/com/festa/config/DatabaseConfig.java:
```diff
@@ -0,0 +1,48 @@
+        sqlSessionFactoryBean.setTypeAliasesPackage("com.festa.dto");
+        sqlSessionFactoryBean.setMapperLocations(new PathMatchingResourcePatternResolver().getResources("classpath:/mappers/*.xml"));        // Mapper 파일 위치 셋팅
+        SqlSessionFactory sqlFactory = sqlSessionFactoryBean.getObject();
+        sqlFactory.getConfiguration().setMapUnderscoreToCamelCase(true);
```
여기서 설정하신 각 옵션들에 대한 의도를 주석으로 남겨주시면 좋을 것 같습니다~

f-lab-dev commented on src/main/java/com/festa/config/JasyptConfig.java:
```diff
@@ -0,0 +1,26 @@
+    public StringEncryptor stringEncryptor() {
+        PooledPBEStringEncryptor encryptor = new PooledPBEStringEncryptor();
+        SimpleStringPBEConfig config = new SimpleStringPBEConfig();
+        config.setPassword("KEY"); // 암호화, 복호화에 사용할 키
```
이 key는 필요에 따라 달라질 것 같은데요. `@Value` 어노테이션을 이용해 외부에서 주입받도록 해주면 어떨까요?
스프링에서 외부 설정을 주입받는 것에 대해서도 알아보시면 좋을 것 같네요

Yo0oN replied:
@Value("${키}") 를 이용하여 application.properties에 있는 설정 값을 읽어올 수 있습니다. 외부에서 주입받는다면, 값이 변하더라도 컴파일 하지 않고 바뀐 값을 적용할 수 있다는 장점이 있습니다.

f-lab-dev commented on src/main/java/com/festa/controller/MemberController.java:
```diff
@@ -1,32 +1,38 @@
+/*
+Controller : View 반환
+            Controller -> View -> DispatcherServlet -> Response
+RestController : Controller + ResponseBody
```
이런 세세한 설명 좋습니다~👍

f-lab-dev commented on src/main/java/com/festa/dto/SignUpDTO.java:
```diff
@@ -0,0 +1,47 @@
+    @NotEmpty : null, "" 불가능, " " 가능
+    @NotBlank : null, "", " " 불가능
+     */
+    @NotBlank(message = "이름을 입력해주세요.")
```
이런 꼼꼼한 유효성 검사도 좋습니다 👍

f-lab-dev commented on src/main/java/com/festa/exception/DuplicatedIDException.java:
```diff
@@ -0,0 +1,7 @@
+package com.festa.exception;
+
+public class DuplicatedIDException extends RuntimeException {
```
중복된 ID를 사용자가 입력했을 때 나는 예외 같은데 `IllegalArgumentException`을 상속받도록 하면 어떨까요?

Yo0oN replied:
IllegalArgumentException은 부적합한 매개변수를 메서드에 전달했을 때 발상하는 예외입니다. 적절한 이메일을 입력하지 않아 발생하는 예외이므로 RuntimeException보다 구체적이고 상황에 맞는것같습니다. 수정하겠습니다.

f-lab-dev commented on src/main/resources/mappers/SignUpMember.xml:
```diff
@@ -1,17 +1,29 @@
     </insert>
 
-    <select id="checkMemberID" parameterType="String">
+    <select id="validateID" parameterType="String">
```
1. 여기서는 유저의 id가 존재하는지 확인하는 쿼리 같은데요. 이 쿼리에 유효성검사 로직이 없는데도 `validate`라는 이름이 사용되고 있습니다.
2. 특정 데이터가 존재하냐는 함수가 mysql에 존재합니다. 해당 함수를 적용해보시면 좋겠습니다~

Yo0oN replied:
EXISTS(서브쿼리) 함수를 다시 적용시켰습니다. 해당 함수는 결과가 한 건이라도 존재하면  TRUE를, 없으면 FALSE를 리턴합니다.
또한, 일치하는 결과가 발생하면 쿼리를 더이상 수행하지 않습니다.

Yo0oN committed:
feat: 회원가입 기능

exists 함수를 사용하도록 쿼리 수정
중복 예외 시 부적합한 매개변수를 전달 시 발생하는 IllegalArgumentException 상속으로 수정
설정값 외부 주입으로 변경

[#2] 회원가입 기능

Yo0oN committed:
Merge branch 'feature/#2/yj' of https://github.com/f-lab-edu/event-recommender-festa into feature/#2/yj

Yo0oN committed:
feat: 회원가입 기능

userID 필드 추가
userID 존재 여부 메서드 추가
email 중복 여부가 아닌 존재 여부로 수정

[#2] 회원가입 기능

Jane096 commented on src/main/java/com/festa/service/AccountsService.java:
```diff
@@ -17,33 +15,44 @@
+     *
+     * @param userID
+     */
+    public boolean existedID(String userID) {
+        boolean existedID = accountsDAO.existedID(userID);
+        return existedID;
```
이 부분은
```java
return accountsDAO.existedID(userID);
```
라고 한줄로 쓸 수 있을 것 같아요!

Yo0oN replied:
네 수정하겠습니다!

Yo0oN committed:
[#6]feat: 로그인 기능

session 추가 전 로그인 기능 구현

[#6] 로그인 기능

Yo0oN committed:
[#6]feat: 로그인 기능

session 추가

[#6] 로그인 기능

f-lab-dev commented on src/main/java/com/festa/config/DatabaseConfig.java:
```diff
@@ -24,19 +25,17 @@ public HikariConfig hikariConfig() { // application.properties에 작성된 설
-        sqlSessionFactoryBean.setMapperLocations(new PathMatchingResourcePatternResolver().getResources("classpath:/mappers/*.xml"));        // Mapper 파일 위치 셋팅
-        SqlSessionFactory sqlFactory = sqlSessionFactoryBean.getObject();
-        sqlFactory.getConfiguration().setMapUnderscoreToCamelCase(true);
+        sqlSessionFactoryBean.setVfs(SpringBootVFS.class); // 스프링부트 가상파일 시스템을 설정
```
이거를 설정해줘야하는 이유가 무엇일까요?

f-lab-dev commented on src/main/java/com/festa/config/JasyptConfig.java:
```diff
@@ -9,12 +9,17 @@
 @Configuration
 public class JasyptConfig {
+    /*
+    @Value : application.properties에 있는 설정 값을 불러올 수 있다.
```
`application.properties` 말고도 다른 property들도 불러올 수 있습니다~
`spring property`로 검색해보시면 좋을 것 같네요

Yo0oN replied:
스프링부트는 application.properties를 기본적으로 읽어오는데, 그 외의 설정파일을 읽어오고 싶다면 @PropertySource를 이용하여 읽어오고, @Value를 이용하여 값을 사용할 수 있습니다!

f-lab-dev commented on src/main/resources/mappers/SignUpMember.xml:
```diff
@@ -21,11 +21,13 @@
-        FROM USERINFO
-        WHERE
-        email = #{email}
+    <select id="validateID" parameterType="boolean">
```
이 쿼리의 역할은 해당 email을 가진 유저가 존재하는지를 검색하는 쿼리같습니다. 이 이름이 그 역할을 잘 나타내고있다고 볼 수 있을까요?

Yo0oN replied:
이부분은 현재 existedEmail로 수정되어있는데 한번 더 수정해서 return값이 boolean이기 때문에 isValidatedID로 수정하겠습니다!

f-lab-dev commented on src/main/java/com/festa/dao/AccountsDAO.java:
```diff
@@ -9,5 +9,7 @@
     int signUp(SignUpDTO signUpDTO);
 
-    boolean validateID(String id);
+    boolean existedEmail(String email);
```
보통 boolean을 리턴하는 메소드의 이름 앞에는 `is`를 붙여줍니다~

Yo0oN replied:
네 수정하겠습니다~

f-lab-dev commented on src/main/java/com/festa/dto/SignUpDTO.java:
```diff
@@ -10,6 +10,7 @@
  setter가 있다면 객체가 언제, 어디서나 변경 가능하기 때문에 일관성을 유지하기 힘들다.
  대신 @AllArgsConstructor 모든 멤버를 매개변수로 받는 생성자를 만들 수 있다.
  */
+@Value
```
이 Value는 어떤 역할을 해줄까요?

Yo0oN replied:
value를 사용하면 기본적으로 private과 final이 붙은 상수가 됩니다. 또 equals, hashCode, toString을 자동으로 만들어줍니다. Data 어노테이션과 비슷하지만 대신 불변객체로 만들 수 있습니다.
불변객체는 상태를 바꿀 수 없는 객체로, 멀티 스레드 환경에서 Thread safe하며, 데이터의 신뢰도가 높아진다는 장점이 있습니다.

Yo0oN committed:
[#6]feat: 로그인 기능

SessionWriter -> SessionManagement 수정
메서드, 쿼리, 클래스 이름 알맞게 변경

[#6] 로그인 기능

Yo0oN committed:
[#6]feat: 로그인 기능

session 확인하는 부분 조건 오타 수정

[#6] 로그인 기능

Yo0oN committed:
[#6]feat: 로그인 기능

쿼리 이름 수정

[#6] 로그인 기능

f-lab-dev committed:
Merge pull request #10 from f-lab-edu/feature/#6/yj

[#6] feat: 로그인 기능

Yo0oN committed:
[#2]feat: UserLevel, SessionManagement 수정

1. UserLevel enum으로 변경
2. SessionManagement -> SessionManager 이름 변경

[#2] 회원가입, 로그인 기능

f-lab-dev commented on src/main/java/com/festa/service/AccountsService.java:
```diff
@@ -62,7 +62,7 @@ public void login(LoginDTO loginDTO, HttpSession httpSession){
         LoginDTO resultLoginDTO = accountsDAO.getUserInfoForLogin(loginDTO);
 
-        sessionManagement.makeSession(httpSession, loginSessionKey, resultLoginDTO);
+        sessionManager.makeSession(httpSession, loginSessionKey, resultLoginDTO);
```
`makeSession`이름은 **세션을 만들다**로 읽히는데 이 로직은 세션에 값을 삽입하는 로직으로 보입니다.
그리고 이 `SessionManager` 클래스가 꼭 필요한지도 생각해봐야겠네요~

f-lab-dev commented:
이제 사용하지 않는 PR이라면 모두 닫아주세요~
